service/worker/scheduler/workflow.go

```diff
@@ -365,27 +377,76 @@ func (s *scheduler) processPatch(patch *schedpb.SchedulePatch) {
 	}
 }
 
-func (s *scheduler) getNextTime(after time.Time) getNextTimeResult {
-
+func (s *scheduler) getNextTimeV1(after time.Time) getNextTimeResult {
 	// we populate the map sequentially, if after is not in the map, it means we either exhausted
 	// all items, or we jumped through time (forward or backward), in either case, refresh the cache
-	next, ok := s.nextTimeResultCache[after]
+	next, ok := s.nextTimeCacheV1[after]
 	if ok {
 		return next
 	}
-	s.nextTimeResultCache = nil
+	s.nextTimeCacheV1 = nil
 	// Run this logic in a SideEffect so that we can fix bugs there without breaking
 	// existing schedule workflows.
 	panicIfErr(workflow.SideEffect(s.ctx, func(ctx workflow.Context) interface{} {
 		results := make(map[time.Time]getNextTimeResult)
-		for t := after; !t.IsZero() && len(results) < maxNextTimeResultCacheSize; {
-			next := s.cspec.getNextTime(t)
+		for t := after; !t.IsZero() && len(results) < nextTimeCacheV1Size; {
+			next := s.cspec.getNextTime(s.jitterSeed(), t)
 			results[t] = next
 			t = next.Next
 		}
 		return results
-	}).Get(&s.nextTimeResultCache))
-	return s.nextTimeResultCache[after]
+	}).Get(&s.nextTimeCacheV1))
+	return s.nextTimeCacheV1[after]
+}
+
+func (s *scheduler) getNextTimeV2(after time.Time) getNextTimeResult {
+	// Asking for a time before the cache, need to refill
+	if after.Before(s.nextTimeCacheV2.Start) {
+		s.fillNextTimeCacheV2(after)
+	}
+	// Results covers a contiguous time range so we can search in it
+	for _, result := range s.nextTimeCacheV2.Results {
+		if result.Next.After(after) {
+			return result
+		}
+	}
+	// Ran off end: if completed, then we're done
+	if s.nextTimeCacheV2.Completed {
+		return getNextTimeResult{}
+	}
+	// Otherwise refill and take first result, or none
+	s.fillNextTimeCacheV2(after)
+	if len(s.nextTimeCacheV2.Results) > 0 {
+		return s.nextTimeCacheV2.Results[0]
+	}
+	return getNextTimeResult{}
+}
+
+func (s *scheduler) fillNextTimeCacheV2(start time.Time) {
+	// Clear value so we can Get into it
+	s.nextTimeCacheV2.clear()
+	// Run this logic in a SideEffect so that we can fix bugs there without breaking
+	// existing schedule workflows.
+	panicIfErr(workflow.SideEffect(s.ctx, func(ctx workflow.Context) interface{} {
+		cache := nextTimeCacheV2{Start: start}
+		for t := start; len(cache.Results) < s.tweakables.NextTimeCacheV2Size; {
+			next := s.cspec.getNextTime(s.jitterSeed(), t)
+			if next.Next.IsZero() {
```

Comment:
Why do we do this?

Reply:
not all schedules go off to infinity, some end at a certain point and there are no more matching times. e.g. if you use EndTime or if you use a calendar spec with a year match